pkg/sshd/lock.go

```diff
@@ -17,38 +17,56 @@ type RepositoryLock interface {
 	// Lock acquires a lock for a repository. In the case the repository is already locked
 	// it waits until a timeout to get the lock. If it was not possible to get the
 	// lock after the timeout an error is returned.
-	Lock(repoName string, timeout time.Duration) error
+	Lock(repoName string) error
 	// Unlock releases the lock for a repository or returns an error if the specified
 	// name doesn't exist. In the case the repository is already locked it waits until
 	// a timeout to get the lock. If it was not possible to get the lock after the timeout
 	// an error is returned.
-	Unlock(repoName string, timeout time.Duration) error
+	Unlock(repoName string) error
+	// Timeout returns the time duration for which it has to hold the lock
+	Timeout() time.Duration
 }
 
-func wrapInLock(lck RepositoryLock, repoName string, timeout time.Duration, fn func() error) error {
-	if err := lck.Lock(repoName, timeout); err != nil {
+func wrapInLock(lck RepositoryLock, repoName string, fn func() error) error {
+	if err := lck.Lock(repoName); err != nil {
 		return errAlreadyLocked
 	}
-	defer lck.Unlock(repoName, timeout)
-	return fn()
+	timer := time.NewTimer(lck.Timeout())
+	var fnerr error
+	quit := make(chan bool, 1)
+	go func() {
+		<-timer.C
+		quit <- false
+	}()
+	go func() {
+		fnerr = fn()
+		quit <- timer.Stop()
+	}()
+	quitter := <-quit
+	lck.Unlock(repoName)
+	if !quitter {
+		return fmt.Errorf("%s lock exceeded timout", repoName)
+	}
```

Comment:
I get what this is doing and generally how it works, but it took me a few minutes to figure out _why_, and I think the part that's easy to miss is the buffered channel.

I think to make this code easier to read, it would be helpful to:

1. Use a `select` to receive the timer channel in the same goroutine as `wrapInLock` executes
2. Use a `select` to cancel the goroutine that calls `fn()` if execution times out

I've written an example similar to what I'm referring to [here](https://play.golang.org/p/udDlToS1bJ). I think the reduction of the goroutine and removal of the buffered channel makes it easier to grasp. Let me know thoughts.